fsspec/implementations/tests/test_prefix.py

```diff
@@ -0,0 +1,238 @@
+    try:
+        os.chdir(dirname)
+        for filename, text in d.items():
+            filename = Path(filename)
+
+            if not filename.parent.exists():
+                filename.parent.mkdir(parents=True, exist_ok=True)
+
+            f = open(filename, "w" + mode)
+            try:
+                f.write(text)
+            finally:
+                try:
+                    f.close()
+                except AttributeError:
+                    pass
+
+        yield list(d)
+
+        for filename in d:
+            if os.path.exists(filename):
+                try:
+                    os.remove(filename)
+                except (IOError, OSError):
+                    pass
+    finally:
+        os.chdir(odir)
+
+
+def test_open():
+    with filetexts(csv_files, mode="b"):
+        fs = PrefixFileSystem(prefix="a", fs=fsspec.filesystem("file"))
+        with fs.open("b/c/.test.fakedata.3.csv") as f:
+            assert f.read() == b"a,b\n3,4,5\n"
+
+
+def test_prefix_root():
+    with filetexts(csv_files, mode="b"):
+        fs = PrefixFileSystem(prefix="/", fs=fsspec.filesystem("file"))
+        abs_path_file = os.path.abspath("a/b/c/.test.fakedata.3.csv")
+
+        # Risk double root marker (in path and in prefix)
+        with fs.open(abs_path_file) as f:
+            assert f.read() == b"a,b\n3,4,5\n"
+
+        # no root marker in windows paths
+        if os.name != "nt":
+            # Without root marker
+            with fs.open(abs_path_file[1:]) as f:
+                assert f.read() == b"a,b\n3,4,5\n"
+
+
+def test_cats():
+    with filetexts(csv_files, mode="b"):
+        fs = PrefixFileSystem(prefix=".", fs=fsspec.filesystem("file"))
+        assert fs.cat(".test.fakedata.1.csv") == b"a,b\n" b"1,2\n"
+        out = set(fs.cat([".test.fakedata.1.csv", ".test.fakedata.2.csv"]).values())
+        assert out == {b"a,b\n" b"1,2\n", b"a,b\n" b"3,4\n"}
+        assert fs.cat(".test.fakedata.1.csv", None, None) == b"a,b\n" b"1,2\n"
+        assert fs.cat(".test.fakedata.1.csv", start=1, end=6) == b"a,b\n" b"1,2\n"[1:6]
+        assert fs.cat(".test.fakedata.1.csv", start=-1) == b"a,b\n" b"1,2\n"[-1:]
+        assert (
+            fs.cat(".test.fakedata.1.csv", start=1, end=-2) == b"a,b\n" b"1,2\n"[1:-2]
+        )
+        out = set(
+            fs.cat(
+                [".test.fakedata.1.csv", ".test.fakedata.2.csv"], start=1, end=-1
+            ).values()
+        )
+        assert out == {b"a,b\n" b"1,2\n"[1:-1], b"a,b\n" b"3,4\n"[1:-1]}
+
+
+def test_not_found():
+    fn = "not-a-file"
+    fs = PrefixFileSystem(prefix=".", fs=fsspec.filesystem("file"))
+    with pytest.raises((FileNotFoundError, OSError)):
+        with OpenFile(fs, fn, mode="rb"):
+            pass
+
+
+def test_isfile():
+    fs = PrefixFileSystem(prefix=".", fs=fsspec.filesystem("file"))
+    with filetexts(files, mode="b"):
+        for f in files.keys():
+            assert fs.isfile(f)
+            assert fs.isfile("file://" + f)
```

Comment:
Hm, I don't think PrefixFileSystem should be aware of full urls for the filesystem it is wrapping. E.g. `s3://one/two` will mean very different things depending on whether or not `one` is a bucket or if it is a dir within prefixfs wrapper for s3, so to handle that the same way we handle `file://` here, we would have to know how `s3fs` works inside. What we might want though is something name-addresable, like `fs://myfs/path`, which would understand that `myfs` is a PrefixFileSystem instance and would pass `path` to it, which would then be further resolved.